.github/container/jax_nsys/Analysis.ipynb

```diff
@@ -57,9 +57,99 @@
     "module_df = all_data[\"module\"]\n",
     "compile_df = all_data[\"compile\"]\n",
     "# module_df may contain some entries with ProgramId == -1, which are typically\n",
-    "# autotuner executions. Throw these away for now.\n",
-    "module_df = module_df[module_df[\"ProgramId\"] >= 0]\n",
-    "thunk_df = thunk_df[thunk_df[\"ProgramId\"] >= 0]"
+    "# autotuner executions. Throw these away for now; ProgramId is the first\n",
+    "assert module_df.index.names[0] == thunk_df.index.names[0] == \"ProgramId\"\n",
+    "module_df = module_df.loc[0:]\n",
+    "thunk_df = thunk_df.loc[0:]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "313c81ca-87b7-4930-9e4d-f878f36ac61a",
+   "metadata": {},
+   "source": [
+    "## Data format\n",
+    "\n",
+    "First, look at the high-level format of the profile data frames.\n",
+    "`module_df` has a single row for each XLA module execution, which typically corresponds to a single JITed JAX function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ba11b419-4b68-4b74-8449-687a070ac90a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "module_df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0ddc41d5-e3ad-4cf8-adf0-33cf2f9de538",
+   "metadata": {},
+   "source": [
+    "This data frame has a three-level index:\n",
+    "- `ProgramId` is an integer ID that uniquely identifies the XLA module\n",
+    "- This is the `ProgramExecution`-th execution of the module within the profiles. You may see this starting from 1, not 0, because of the `warmup_removal_heuristics` option passed to `load_profiler_data`.\n",
+    "- `Rank` is used in the MPI sense; it is a global index of the GPU on which the module execution took place, across a (potentially distributed) SPMD run\n",
+    "\n",
+    "The columns are as follows:\n",
+    "- `Name`: the name of the XLA module; this should always be the same for a given `ProgramId`\n",
+    "- `ProjStartNs`: the timestamp of the start of the module execution on the GPU, in nanoseconds\n",
+    "- `ProjDurNs`: the duration of the module execution on the GPU, in nanoseconds\n",
+    "- `OrigStartNs`: the timestamp of the start of the module launch **on the host**, in nanoseconds. *i.e.* `ProjStartNs-OrigStartNs` is something like the launch latency of the first kernel\n",
+    "- `OrigDurNs`: the duration of the module launch **on the host**, in nanoseconds\n",
+    "\n",
+    "The other profile data frame for GPU execution is `thunk_df`, which has a single row for each XLA thunk.\n",
+    "Loosely, each XLA module contains a series of thunks, and each thunk launches a GPU kernel.\n",
+    "In reality, thunks can be nested and may launch multiple kernels, but this data frame still provides the most granular distribution available of GPU execution time across the XLA module:"
```

Comment:
Does the data ever contain both a thunk that has children _and_ one or multiple of its children? How would I only operate on the "leaf" thunks?

Reply:
Answers: "no" and "trivially; those are the only ones present in the frame"

This does mean that you'd need to do some leg work to recover the intermediate thunks, of which the most obvious example is a `while` op:
![Screenshot 2024-07-01 at 15 29 15](https://github.com/NVIDIA/JAX-Toolbox/assets/6459623/43083745-8920-4179-8bf7-c196b9a77d62)
there will be no `while.5.0` rows in the thunk frame.